src/upgrade.py

```diff
@@ -0,0 +1,151 @@
+from typing_extensions import override
+
+if TYPE_CHECKING:
+    from charm import ZooKeeperCharm
+
+logger = logging.getLogger(__name__)
+
+
+class ZooKeeperDependencyModel(BaseModel):
+    """Model for ZooKeeper Operator dependencies."""
+
+    service: DependencyModel
+
+
+class ZooKeeperUpgrade(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "ZooKeeperCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    @property
+    def idle(self) -> bool:
+        """Checks if cluster state is idle.
+
+        Returns:
+            True if cluster state is idle. Otherwise False
+        """
+        return self.cluster_state == "idle"
+
+    @cached_property
+    def client(self) -> ZooKeeperManager:
+        """Cached client manager application for performing ZK commands."""
+        return ZooKeeperManager(
+            hosts=self.charm.cluster.active_hosts,
+            client_port=self.charm.cluster.client_port,
+            username="super",
+            password=self.charm.cluster.passwords[0],
+        )
+
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        self.pre_upgrade_check()
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        default_message = "Pre-upgrade check failed and cannot safely upgrade"
+        try:
+            if not self.client.members_broadcasting or not len(self.client.server_members) == len(
+                self.charm.cluster.peer_units
+            ):
+                raise ClusterNotReadyError(
+                    message=default_message,
+                    cause="Not all application units are connected and broadcasting in the quorum",
+                )
+
+            if self.client.members_syncing:
+                raise ClusterNotReadyError(
+                    message=default_message, cause="Some quorum members are syncing data"
+                )
+
+            if not self.charm.cluster.stable:
+                raise ClusterNotReadyError(
+                    message=default_message, cause="Charm has not finished initialising"
+                )
+
+        except QuorumLeaderNotFoundError:
+            raise ClusterNotReadyError(message=default_message, cause="Quorum leader not found")
+        except ConnectionClosedError:
+            raise ClusterNotReadyError(
+                message=default_message, cause="Unable to connect to the cluster"
+            )
+        except Exception as e:
+            logger.debug(str(e))
+            raise ClusterNotReadyError(message=default_message, cause="Unknown error")
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        for unit in self.charm.cluster.peer_units:
+            config = self.charm.cluster.unit_config(unit=unit)
+
+            # upgrade quorum leader last
```

Comment:
question: Is this something explicitly needed for ZK?

Reply:
Yup, not so important for Kafka as every unit is a 'leader' of a partition.